be/src/http/action/tablet_migration_action.cpp

```diff
@@ -0,0 +1,258 @@
+    const std::string& req_tablet_id = req->param("tablet_id");
+    const std::string& req_schema_hash = req->param("schema_hash");
+    try {
+        tablet_id = std::stoull(req_tablet_id);
+        schema_hash = std::stoul(req_schema_hash);
+    } catch (const std::exception& e) {
+        LOG(WARNING) << "invalid argument.tablet_id:" << req_tablet_id
+                     << ", schema_hash:" << req_schema_hash;
+        return Status::InternalError(strings::Substitute("Convert failed, $0", e.what()));
+    }
+    dest_disk = req->param("disk");
+    goal = req->param("goal");
+    if (goal != "run" && goal != "status") {
+        return Status::InternalError(strings::Substitute("invalid goal argument."));
+    }
+    return Status::OK();
+}
+
+Status TabletMigrationAction::_check_migrate_request(int64_t tablet_id, int32_t schema_hash,
+                                                     string dest_disk, TabletSharedPtr& tablet,
+                                                     DataDir** dest_store) {
+    tablet = StorageEngine::instance()->tablet_manager()->get_tablet(tablet_id, schema_hash);
+    if (tablet == nullptr) {
+        LOG(WARNING) << "no tablet for tablet_id:" << tablet_id << " schema hash:" << schema_hash;
+        return Status::NotFound("Tablet not found");
+    }
+
+    // request specify the data dir
+    *dest_store = StorageEngine::instance()->get_store(dest_disk);
+    if (*dest_store == nullptr) {
+        LOG(WARNING) << "data dir not found: " << dest_disk;
+        return Status::NotFound("Disk not found");
+    }
+
+    if (tablet->data_dir() == *dest_store) {
+        LOG(WARNING) << "tablet already exist in destine disk: " << dest_disk;
+        return Status::AlreadyExist("Tablet already exist in destination disk");
+    }
+
+    // check disk capacity
+    int64_t tablet_size = tablet->tablet_footprint();
+    if ((*dest_store)->reach_capacity_limit(tablet_size)) {
+        LOG(WARNING) << "reach the capacity limit of path: " << (*dest_store)->path()
+                     << ", tablet size: " << tablet_size;
+        return Status::InternalError("Insufficient disk capacity");
+    }
+
+    return Status::OK();
+}
+
+Status TabletMigrationAction::_execute_tablet_migration(TabletSharedPtr tablet,
+                                                        DataDir* dest_store) {
+    int64_t tablet_id = tablet->tablet_id();
+    int32_t schema_hash = tablet->schema_hash();
+    string dest_disk = dest_store->path();
+    EngineStorageMigrationTask engine_task(tablet, dest_store);
+    OLAPStatus res = StorageEngine::instance()->execute_task(&engine_task);
+    Status status = Status::OK();
+    if (res != OLAP_SUCCESS) {
+        LOG(WARNING) << "tablet migrate failed. tablet_id=" << tablet_id
+                     << ", schema_hash=" << schema_hash << ", dest_disk=" << dest_disk
+                     << ", status:" << res;
+        status = Status::InternalError("migration task failed");
```

Comment:
better to add `res` to the message in `status`, so that user can known what's wrong

Reply:
> better to add `res` to the message in `status`, so that user can known what's wrong

That sounds reasonable.